src/main/java/io/split/android/client/cache/SplitCache.java

```diff
@@ -65,34 +83,81 @@ public boolean setChangeNumber(long changeNumber) {
 
     @Override
     public long getChangeNumber() {
+        return mChangeNumber;
+    }
+
+    @Override
+    public Split getSplit(String splitName) {
+        Split split =  mInMemorySplits.get(splitName);
+        if(split == null && !mRemovedSplits.contains(splitName)) {
+            split = getSplitFromDisk(splitName);
+            mInMemorySplits.put(splitName, split);
+        }
+        return split;
+    }
+
+    @Override
+    public List<String> getSplitNames() {
+        return new ArrayList(mInMemorySplits.keySet()) ;
+    }
+
+    private Split getSplitFromDisk(String splitName){
+        Split split = null;
+
+        if(Strings.isNullOrEmpty(splitName)) {
+            return null;
+        }
+
         try {
-            return Long.parseLong(_storage.read(getChangeNumberId()));
+            String splitJson = mFileStorageManager.read(getSplitId(splitName));
+            if (splitJson != null && !splitJson.trim().equals("")) {
+                split = Json.fromJson(splitJson, Split.class);
+            }
+        } catch (IOException e) {
+            Logger.e(e, "Unable to load split from disk");
+        } catch (JsonSyntaxException syntaxException) {
+            Logger.e(syntaxException, "Unable to parse saved segments");
```

Comment:
If thi shappens... do we do any effort to clean the malformed split from disk? I mean do we rewrite it or delete it? If not this will happen forever

Reply:
It will only be rewrote if a new split change (for that item) is downloaded.
Now when json parsing fails the file is removed.